src/Build/Instance/TaskRegistry.cs

```diff
@@ -1794,6 +1788,8 @@ public void Translate(ITranslator translator)
             if (translator.Mode == TranslationDirection.ReadFromStream)
             {
                 _taskRegistrations = (Dictionary<RegisteredTaskIdentity, List<RegisteredTaskRecord>>)copy;
+                // Ensure that mutations of the deserialized task registry are getting unique order ids.
+                s_nextRegistrationOrderId = Math.Max(s_nextRegistrationOrderId, copy?.Count ?? 0);
```

Comment:
I'm still not convinced this is correct.
1. If multiple threads race to update `s_nextRegistrationOrderId`, the value we end up may be smaller than `copy?.Count`. Could be fixed with a lock or an `Interlocked.CompareExchange` loop.
2. In https://github.com/dotnet/msbuild/pull/8861#discussion_r1232155528 you wrote that we assume tasks are not moved between processes. Are you sure this is true? It looks like `ProjectInstance` can be sent to an OOP node and if `_translateEntireState` is true, the task registry is moved as well.

Reply:
Done
@ladipro - do you want to have a second look?
I've made the sequence counter an instance of TaskRegistry and made it part of the serialization payload as you've originaly suggested on the previous PR - it seems to be the least confusing and most future proof option.
Btw. the `Interlocked` is not even needed - as the access to `RegisterTask` is serialized (if it would not, then there is no guarantee that the order of registration calls is same as the actual order of the instantiations). But I'm keeping it as it's more selfdescriptive. But very open to oposing thoughts.

Reply:
Thank you, I like it!